src/lib/genesis_ledger_helper/genesis_ledger_helper.ml

```diff
@@ -107,25 +107,85 @@ module Accounts = struct
     let%map pk = Signature_lib.Public_key.Compressed.gen in
     (None, Account.create (Account_id.create pk Token_id.default) balance)
 
+  let gen : (Private_key.t option * Account.t) Quickcheck.Generator.t =
+    let open Quickcheck.Generator.Let_syntax in
+    let%bind balance = Int.gen_incl 10 500 >>| Currency.Balance.of_int in
+    gen_with_balance balance
+
   let generate n : (Private_key.t option * Account.t) list =
     let open Quickcheck in
     random_value ~seed:(`Deterministic "fake accounts for genesis ledger")
       (Generator.list_with_length n gen)
 
-  let rec pad_to n accounts =
+  (* This implements a tail-recursive generator using the low-level primitives
+     so that we don't blow out the stack.
+  *)
+  let gen_balances_rev balances :
+      (Private_key.t option * Account.t) list Quickcheck.Generator.t =
+    match balances with
+    | [] ->
+        Quickcheck.Generator.return []
+    | (n, balance) :: balances_tl ->
+        Quickcheck.Generator.create (fun ~size ~random ->
+            let rec gen_balances_rev n balance balances_tl accounts =
+              if n > 0 then
+                let new_random = Splittable_random.State.split random in
+                let account =
+                  (* Manually generate an account using the [generate] primitive. *)
+                  Quickcheck.Generator.generate ~size ~random:new_random
+                    (gen_with_balance balance)
+                in
+                gen_balances_rev (n - 1) balance balances_tl
+                  (account :: accounts)
+              else
+                match balances_tl with
+                | [] ->
+                    accounts
+                | (n, balance) :: balances_tl ->
+                    gen_balances_rev n balance balances_tl accounts
+            in
+            gen_balances_rev n balance balances_tl [] )
+
+  let pad_with_rev_balances balances accounts =
+    let balances_accounts =
+      Quickcheck.random_value
+        ~seed:(`Deterministic "fake accounts with balances for genesis ledger")
+        (gen_balances_rev balances)
+    in
+    List.append accounts balances_accounts
+
+  (* NOTE: When modifying this function, be very careful to ensure that all
+     operations are tail-recursive, otherwise a sufficiently large genesis
+     ledger will blow the stack.
+     In particular, do not use any functions that return values of the form
+     [_ :: _], since this construction is NOT tail-recursive.
+  *)
+  let pad_to n accounts =
     if n <= 0 then accounts
     else
-      match accounts with
-      | [] ->
-          generate n
-      | account :: accounts ->
-          account :: pad_to (n - 1) accounts
+      let exception Stop in
+      try
+        (* Count accounts and reverse the list while we're doing so to avoid
+           re-traversing the list.
+        *)
+        let rev_accounts, count =
+          List.fold ~init:([], 0) accounts ~f:(fun (acc, count) account ->
```

Comment:
could use `List.foldi` here, instead of rolling your own

Reply:
I don't this gains us anything except an extra redundant argument in the iterator:

```ocaml
List.foldi ~init:([], 0) accounts ~f:(fun count (acc, _also_count) account ->
  let count = count + 1 in
  if count >= n then raise Stop ;
  (account :: acc, count + 1) )
```